### ISSUE TITLE
Detection scheme needed for v0.7 mcHF boards

### ISSUE THREAD

**Issue by frazierjason** [df8oe/mchf-github#1365]
I commented in issue #1319 that:

> one difference in the v0.7 schematic that was not present in v0.6: a 4.7k pullup resistor on I/O pin PB2 (BOOT0) that goes to the BAND+ button situated across the anti-spark capacitor. Were other versions prior to v0.6 also lacking that 4.7k pullup resistor?
I wonder if we could do any kind of trick to detect that pullup resistor, which would tell us we're on a v0.7.

@df8oe replied:

> This can be used to identify 0.7 boards. During boot we can add a pulldown resistor to BOOT0 and check if level at PB2 is low. If it is, we are on older revisions, if not, we are on 0.7. Possible issue can be that external resistor is to high so that level does not increase enough - but simple replacing (e.g. with 1K) would solve that.

This probably needs to be pulled out to its own issue here and worked separately.  I'd like to take a crack at it, but I might be in over my head at this time.  For sure I'd need some guidance points:

- Where to store the v0.7 board model once anyone has the detection designed?  I see there are 0x29 signatures in SRAM2 for rfmod/special boot conditions.  What SRAM2+?? address offset should we define to store a signature indicating mcHF revision?  I presume we'd just increment it simply, say 0x26 for v0.6 and lower (equivalence them all), and 0x27 for v0.7.  I picked these arbitrarily, not sure why 0x29 was used as signatures in other SRAM2 offsets.
- If using the weak internal pulldown resistor (30 to 50k) doesn't work out, we could possibly use the antistatic 100pF capacitor there together with the external 4.7k resistor as an RC circuit and detect them.  If we put PB2 into output mode and set it low, that should draw only 0.7 mA current through the external pullup resistor, and we just hold it there long enough to drain the antistatic capacitor.  Then we set it back to input mode, and we can only hope the pin stays low upon flipping to input mode.  Then we just measure the rise time until we get to logic high -- there should be a longer wait time only for v0.7.  Total rise time for such an RC circuit is about 1 millisecond, so it should be detectable if we are not interrupted during measurement.
- If I am totally off on these ideas, please feel free to calibrate me.  We all learn by doing.

**Comment by df8oe** [comment]
Don't think about any SRAM2 stuff. That is only for very, very special usages like exchanging varaibles between bootloader and firmwware.

You can/shuold use a element in tc-structure.

Regarding detection I can give you a hint - but cannot do ist by myself at this time beause I am very busy with OVI40.

At starting up do the following:
1) initialize GPIO as "input" with "pulldown" resistor
2) check level at the GPIO. If it is HIGH you are on 0.7 board. If it is low you are on older boards. Save this state in a ts-structure variable and use it later for swapping directions
3) re-initilize the GPIO to pullup resistor

73
Andreas

**Comment by frazierjason** [comment]
I've tried so many combinations of reading the BAND+ (PB2) input by now, and none of them yield any difference whatsoever between mcHF v0.6 vs 0.7.

I know I'm reading PB2 correctly, because I temporarily disabled the button test method, and my input checking logic sees the PB2 pulled low if I hold down the BAND+ button just shortly after powering on the unit, only when I have the internal pullup configured.  I am validating my logic by changing the color of the PF1 label when not in the menu, since I don't have debug print enabled (yet).  In this manner I sped up my tests by trying three combinations at a time, and coloring PF1, PF2, PF3 on boot.  It's still slow going, I need to compile and reflash for every new set of tries.

Besides checking PB2 as input with internal pulldown, I've tried taking advantage of the 100pF capacitor acting with the external 4.7k resistor as an RC circuit, by quickly switching the ~40k internal from pullup to pulldown, or pulldown to pullup, or tossing in a nopull state, in all the possible combinations, plus all of these again while separating them with a really dumb while() loop to decrement a number as a form of wait to let the RC circuit (dis-)charge.  I also checked the pin state before *and* after changing the internal resistor state, there is never any delta detected.  No matter what I do, the combined external and internal pull up/down resistance is just not enough to drag that line to a low enough voltage -- PB0 always reads in the set state.

I'm wondering, would it be dangerous to first briefly set PB2 as an open drain or push-pull output line, so that the ~0.7mA current flowing through the external resistor is sunk mostly by PB2 rather than to the ground at the other side of the 100pF capacitor? Would this allow the capacitor across the switch to fully bleed off?  Then when we flip back to input mode, and PB2 voltage is rising, initially the RC reaction might stay low for long enough to get detected as reset state?  I feel like I've got this all wrong, and grasping at straws.  And I don't know enough about how the load is represented inside of PB2 while in input mode.

**Comment by df8oe** [comment]
I think value of external pullup resistor is much too low. try to change it e.g. to 100k, 150K, 220k. Your internally added pulldown must be able to initiate a "low" state on that GPIO. If this works all others have to swap the external pullup ("a new mod"). Hopefully Chris notices that and changes his original design...

**Comment by frazierjason** [comment]
I tried a few more last effort detection schemes, including starting with output open drain mode, then switching to input mode, these didn't yield any new differences.  I also tried directly setting MODER and PUPDR rather than using HAL_GPIO_INIT just for the flip between modes/internal resistor settings, to cut down on clock cycles since there's only a few hundred nanoseconds of RC rise/fall time on PB2.  None of them worked, and I have a nagging feeling that a clone radio could come along and ruin my detection scheme anyway, even if it worked.

So I am giving up on mcHF v0.7 detection, and will instead settle for a user-configurable option that I will put in the Display Settings.
